hyperlink/_url.py

```diff
@@ -1464,3 +1504,329 @@ def remove(self, name):
+
+        Args:
+          text (unicode): Text containing the URL
+          lazy (bool): Whether to pre-decode all parts of the URL to
+              check for validity. Defaults to True.
+        """
+        _url = URL.from_text(text)
+        return cls(_url, lazy=lazy)
+
+    @property
+    def encoded_url(self):
+        """Access the underlying :class:`URL` object, which has any special
+        characters encoded.
+        """
+        return self._url
+
+    def to_text(self, *a, **kw):
+        "Passthrough to :meth:`~hyperlink.URL.to_text()`"
+        return self._url.to_text(*a, **kw)
+
+    def to_uri(self, *a, **kw):
+        "Passthrough to :meth:`~hyperlink.URL.to_uri()`"
+        return self._url.to_uri(*a, **kw)
+
+    def to_iri(self, *a, **kw):
+        "Passthrough to :meth:`~hyperlink.URL.to_iri()`"
+        return self._url.to_iri(*a, **kw)
+
+    def click(self, href=u''):
+        "Return a new DecodedURL wrapping the result of :meth:`~hyperlink.URL.click()`"
+        return type(self)(self._url.click(href=href))
+
+    def sibling(self, segment):
+        """Automatically encode any reserved characters in *segment* and
+        return a new DecodedURL wrapping the result of
+        :meth:`~hyperlink.URL.sibling()`
+        """
+        return type(self)(self._url.sibling(_encode_reserved(segment)))
+
+    def child(self, *segments):
+        """Automatically encode any reserved characters in *segments* and
+        return a new DecodedURL wrapping the result of
+        :meth:`~hyperlink.URL.child()`.
+        """
+        if not segments:
+            return self
+        new_segs = [_encode_reserved(s) for s in segments]
+        return type(self)(self._url.child(*new_segs))
+
+    def normalize(self, *a, **kw):
+        "Return a new DecodedURL wrapping the result of :meth:`~hyperlink.URL.normalize()`"
+        return type(self)(self._url.normalize(*a, **kw))
+
+    @property
+    def absolute(self):
+        return self._url.absolute
+
+    @property
+    def scheme(self):
+        return self._url.scheme
+
+    @property
+    def host(self):
+        return _decode_host(self._url.host)
+
+    @property
+    def port(self):
+        return self._url.port
+
+    @property
+    def rooted(self):
+        return self._url.rooted
+
+    @property
+    def path(self):
+        try:
+            return self._path
+        except AttributeError:
+            pass
+        self._path = tuple([_percent_decode(_encode_path_part(p),
+                                            raise_subencoding_exc=True)
+                            for p in self._url.path])
+        return self._path
+
+    @property
+    def query(self):
+        try:
+            return self._query
+        except AttributeError:
+            pass
+        _q = [tuple(_percent_decode(_encode_query_part(x),
+                                    raise_subencoding_exc=True)
+                    if x is not None else None
```

Comment:
Can I have a key of `None` with a non-`None` value?  The `_textcheck` for the query parameter keys in `URL.__init__` isn't nullable, so I think the answer is never.  A defensive change would only check the value for `None`.